### PR TITLE
Implement tensor weak references

### PR DESCRIPTION
**Pull request by apaszke**
Add `WeakTensor` - a `Tensor` counterpart which doesn't keep the data (or any other expensive resources) alive. They can be `.lock()`ed and return `at::optional<Tensor>` if they're still alive.

### CODE REVIEW

**Comment by apaszke** [comment]
@zdevito

**Comment by ezyang** [comment]
CC @smessmer This is diverging further from your implementation in C10.

**Review by ezyang — APPROVED** [review]

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/templates/Tensor.h [hunk 1]
```diff
@@ -198,6 +200,46 @@ struct Tensor : public detail::TensorBase {
...
+  }
+
+  // non-retaining
+  TensorImpl * unsafeGetTensorImpl() const {
+    return pImpl;
+  }
+
+  // XXX: this can return undefined tensors
+  // Ideally it would be at::optional<Tensor>, but MSVC is too cool for that
+  Tensor lock() {
```
Comment (colesbury) [inline aten/src/ATen/templates/Tensor.h]: This should be `const`, right?

https://en.cppreference.com/w/cpp/memory/weak_ptr/lock

File: aten/src/ATen/templates/Tensor.h [hunk 2]
```diff
@@ -198,6 +200,46 @@ struct Tensor : public detail::TensorBase {
...
+
+struct WeakTensor : public detail::WeakTensorBase {
+  using WeakTensorBase = detail::WeakTensorBase;
+  WeakTensor() : WeakTensorBase() {}
+  WeakTensor(TensorImpl * self, bool retain) : WeakTensorBase(self, retain) {}
+  WeakTensor(const WeakTensor & rhs) = default;
+  WeakTensor(WeakTensor && rhs) noexcept = default;
+  WeakTensor(const Tensor& t) : WeakTensorBase(t.pImpl, true) {}
+
+  // reimplemented from TensorBase so the return type is Tensor rather than TensorBase
```
Comment (colesbury) [inline aten/src/ATen/templates/Tensor.h]: "return type is Tensor" -> "return type is WeakTensor"

File: aten/src/ATen/Retainable.h [hunk 3]
```diff
@@ -7,21 +7,52 @@ namespace at {
...
+      // all the decrements and releaseResources().
+      if (weak_refcount == 1) {
+        delete this;
+      } else {
+        releaseResources();
+        weakRelease();
+      }
+    }
+  }
+  void weakRetain() {
```
[the remark this answers is not in the document]
Comment (apaszke) [reply aten/src/ATen/Retainable.h]: alright, I've been trying to match the rest of ATen (e.g. `unsafeGetTensorImpl()`) but I can change it if you want

File: aten/src/ATen/Retainable.h [hunk 4]
```diff
@@ -7,21 +7,52 @@ namespace at {
 // base class for refcounted things, allows for collects of generic
 // refcounted objects that include tensors
 struct Retainable {
-  Retainable(): refcount(1) {}
+  Retainable(): refcount(1), weak_refcount(1) {}
```
Comment (dzhulgakov) [inline aten/src/ATen/Retainable.h]: would it cause perf overhead as Retainable is used in several places and it increases the size of struct?

Maybe I'm missing the context, but what is the motivation for this PR? It diverges quite a bit from the unified design we were hashing out. @ezyang @smessmer

Comment (ezyang) [reply aten/src/ATen/Retainable.h]: The immediate motivation for weak references is for thread-local tracing (https://github.com/pytorch/pytorch/pull/9414). That in and of itself wouldn't be a good reason, but we had a previous case where we needed weak references for a weak mapping of storages to file descriptors in our multiprocessing implementation. And we think that weak references are generally useful enough to deserve the extra word in the tensor struct to hold them.

Comment (apaszke) [reply aten/src/ATen/Retainable.h]: Also, if it really turns out to be too expensive for some reason, we could effectively pack the weak refcount together with regular refcount in a single integer (but I'd really treat that as a last resort).

**Pull request closed without merging**